addons/mail/static/tests/core/record_tests.js

```diff
@@ -149,3 +149,34 @@ QUnit.test("Trusted insert on html field with { html: true }", async (assert) =>
     assert.strictEqual(hello.body.toString(), "<p>hello</p>");
     assert.strictEqual(world.body, "<p>world</p>");
 });
+
+QUnit.test("Unshifting several records preserve their order", async (assert) => {
+    class Child extends Record {
+        static id = "id";
+        id;
+    }
+    class Parent extends Record {
+        static id = "id";
+        id;
+        children = Record.many("Child");
+    }
+    Child.register();
+    Parent.register();
+    const store = await start();
+    const parent = store.Parent.insert({ id: 1 });
+    parent.children.unshift(...[{ id: 3 }, { id: 2 }, { id: 1 }]);
+    assert.deepEqual(
+        parent.children.map((child) => child.id),
+        [3, 2, 1]
+    );
+    parent.children.unshift(...[{ id: 6 }, { id: 5 }, { id: 4 }]);
+    assert.deepEqual(
+        parent.children.map((child) => child.id),
+        [6, 5, 4, 3, 2, 1]
+    );
+    parent.children.unshift({ id: 7 });
+    assert.deepEqual(
+        parent.children.map((child) => child.id),
+        [7, 6, 5, 4, 3, 2, 1]
+    );
+});
```

Comment:
It's easier to understand test with actual business objects, rather than broad & generic `Parent` and `Child`. You can use simplified `Thread` and `Message` models, which actually matches the actual use-case in prod code.

You can take inspiration of previous tests. For example, to simplify readability of test, messages are identified by their content.

Reply:
Also `unshift(...[1, 2, 3])` is equivalent to `unshift(1, 2, 3)`

Reply:
Changed for `Thread`/`Message` as suggested. However I think ids are easier to understand for this test than bodies.